packages/jest-message-util/src/visitor.ts

```diff
@@ -0,0 +1,158 @@
+import {relative} from 'path';
+import {codeFrameColumns} from '@babel/code-frame';
+import {
+  Context,
+  ErrorNode,
+  FrameNode,
+  PathLocatorNode,
+  PrintVisitor,
+  SiteNode,
+  getAbsoluteSitePath,
+} from '@stack-tools/node-tools';
+import chalk = require('chalk');
+import * as fs from 'graceful-fs';
+import micromatch = require('micromatch');
+import type {Config} from '@jest/types';
+
+type Path = Config.Path;
+
+const PATH_NODE_MODULES = `/node_modules/`;
+const PATH_JEST_PACKAGES = `/jest/packages/`;
+const STACK_TRACE_COLOR = chalk.dim;
+const NOT_EMPTY_LINE_REGEXP = /^(?!$)/gm;
+
+const indentLines = (lines: string, indent: string = '  ') =>
+  indent ? lines.replace(NOT_EMPTY_LINE_REGEXP, indent) : lines;
+
+export type FileFrame = FrameNode & {
+  type: 'CallSiteFrame';
+  callSite: {
+    site: {type: 'FileSite'; locator: {type: 'FileLocator' | 'URILocator'}};
+  };
+};
+
+export const getTopFrame = (frames: Array<FrameNode>): FileFrame | null => {
+  return (
+    frames.find((frame): frame is FileFrame => {
+      const absPath = getAbsoluteSitePath(frame);
+
+      return (
+        !!absPath &&
+        !(
+          absPath.includes(PATH_NODE_MODULES) ||
+          absPath.includes(PATH_JEST_PACKAGES)
+        )
+      );
+    }) || null
+  );
+};
+
+export type JestVisitorOptions = {
+  frames: boolean;
+  rootDir: string;
+  testMatch: Array<string>;
+  noCodeFrame: boolean;
+  noStackTrace: boolean;
+  relativeTestPath: Path | null;
+  indentHeader: string;
+  indentFrames: string;
+};
+
+export class JestVisitor<O extends JestVisitorOptions> extends PrintVisitor<O> {
+  constructor(context: Context, options: O) {
+    super(context, options);
+  }
+
+  ErrorBody(error: ErrorNode): string | undefined {
+    const header = this.ErrorHeader(error);
+    const frames = this.Frames(error.frames);
+
+    return frames && header ? `${header}\n\n${frames}` : header || frames;
+  }
+
+  ErrorHeader(error: ErrorNode): string | undefined {
+    const {options} = this;
+    const name = this.visit(error.name);
+    const name_ = name === 'Error' ? undefined : name;
+    const message = error.message ? this.visit(error.message) : '';
+    const message_ = !name && message === 'Error' ? undefined : message;
+    const header =
+      message_ && name_ ? `${name_}: ${message_}` : name_ || message_;
+    return header && indentLines(header, options.indentHeader);
```

Comment:
This logic is messy, but this is exactly the logic needed to get the tests to pass. At least it's all in one place where it is spelled out explicitly.